blockchain/difficulty.go

```diff
@@ -743,19 +748,277 @@ func (b *BlockChain) calcNextRequiredStakeDifficulty(curNode *blockNode) (int64,
+		nextDiff = maximumStakeDiff
+	}
+	if nextDiff < params.MinimumStakeDiff {
+		nextDiff = params.MinimumStakeDiff
+	}
+	return nextDiff
+}
+
+// calcNextRequiredStakeDifficultyV2 calculates the required stake difficulty
+// for the block after the passed previous block node based on the algorithm
+// defined in DCP0001.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) calcNextRequiredStakeDifficultyV2(curNode *blockNode) (int64, error) {
+	// Stake difficulty before any tickets could possibly be purchased is
+	// the minimum value.
+	nextHeight := int64(0)
+	if curNode != nil {
+		nextHeight = curNode.height + 1
+	}
+	stakeDiffStartHeight := int64(b.chainParams.CoinbaseMaturity) + 1
+	if nextHeight < stakeDiffStartHeight {
+		return b.chainParams.MinimumStakeDiff, nil
+	}
+
+	// Return the previous block's difficulty requirements if the next block
+	// is not at a difficulty retarget interval.
+	intervalSize := b.chainParams.StakeDiffWindowSize
+	curDiff := curNode.header.SBits
+	if nextHeight%intervalSize != 0 {
+		return curDiff, nil
+	}
+
+	// Get the pool size and number of tickets that were immature at the
+	// previous retarget interval
+	//
+	// NOTE: Since the stake difficulty must be calculated based on existing
+	// blocks, it is always calcuated for the block after a given block, so
+	// the information for the previous retarget interval must be retrieved
+	// relative to the block just before it to coincide with how it was
+	// originally calculated.
+	var prevPoolSize int64
+	prevRetargetHeight := nextHeight - intervalSize - 1
+	prevRetargetNode, err := b.ancestorNode(curNode, prevRetargetHeight)
+	if err != nil {
+		return 0, err
+	}
+	if prevRetargetNode != nil {
+		prevPoolSize = int64(prevRetargetNode.header.PoolSize)
+	}
+	ticketMaturity := int64(b.chainParams.TicketMaturity)
+	prevImmatureTickets, err := b.sumPurchasedTickets(prevRetargetNode,
+		ticketMaturity)
+	if err != nil {
+		return 0, err
+	}
+
+	// Return the existing ticket price for the first few intervals to avoid
+	// division by zero and encourage initial pool population.
+	prevPoolSizeAll := prevPoolSize + prevImmatureTickets
+	if prevPoolSizeAll == 0 {
+		return curDiff, nil
+	}
+
+	// Count the number of currently immature tickets.
+	immatureTickets, err := b.sumPurchasedTickets(curNode, ticketMaturity)
+	if err != nil {
+		return 0, err
+	}
+
+	// Calculate and return the final next required difficulty.
+	curPoolSizeAll := int64(curNode.header.PoolSize) + immatureTickets
+	return calcNextStakeDiffV2(b.chainParams, nextHeight, curDiff,
+		prevPoolSizeAll, curPoolSizeAll), nil
+}
+
+// sdiffAlgoDeploymentVersion returns the deployment vesion for the stake
+// difficulty algorithm change as defined in DCP0001 for the provided network.
+//
+// This function is safe for concurrent access.
+func sdiffAlgoDeploymentVersion(network wire.CurrencyNet) uint32 {
+	if network != wire.MainNet {
```

Comment:
It shouldn't ever need to be updated except to remove it altogether once the agendas are removed.